client/app/components/permissions-editor/PermissionsEditorDialog.jsx

```diff
@@ -0,0 +1,179 @@
+      </div>
+    </>
+  );
+}
+
+PermissionsEditorDialogHeader.propTypes = { context: PropTypes.oneOf(['query', 'dashboard']) };
+PermissionsEditorDialogHeader.defaultProps = { context: 'query' };
+
+function UserSelect({ onSelect, shouldShowUser }) {
+  const [loadingUsers, setLoadingUsers] = useState(true);
+  const [users, setUsers] = useState([]);
+  const [searchTerm, setSearchTerm] = useState('');
+
+  const debouncedSearchUsers = useCallback(debounce(
+    search => searchUsers(search)
+      .then(setUsers)
+      .finally(() => setLoadingUsers(false)),
+    DEBOUNCE_SEARCH_DURATION,
+  ), []);
+
+  useEffect(() => {
+    setLoadingUsers(true);
+    debouncedSearchUsers(searchTerm);
+  }, [searchTerm]);
+
+  return (
+    <Select
+      className="w-100 m-b-10"
+      placeholder="Add users..."
+      showSearch
+      onSearch={setSearchTerm}
+      suffixIcon={loadingUsers ? <i className="fa fa-spinner fa-pulse" /> : <i className="fa fa-search" />}
+      filterOption={false}
+      notFoundContent={null}
+      value={undefined}
+      getPopupContainer={trigger => trigger.parentNode}
+      onSelect={onSelect}
+    >
+      {users.filter(shouldShowUser).map(user => (
+        <Option key={user.id} value={user.id}>
+          <UserPreviewCard user={user} />
+        </Option>
+      ))}
+    </Select>
+  );
+}
+
+UserSelect.propTypes = {
+  onSelect: PropTypes.func,
+  shouldShowUser: PropTypes.func,
+};
+UserSelect.defaultProps = { onSelect: () => {}, shouldShowUser: () => true };
+
+function PermissionsEditorDialog({ dialog, author, context, aclUrl }) {
+  const [loadingGrantees, setLoadingGrantees] = useState(true);
+  const [grantees, setGrantees] = useState([]);
+  const { loadGrantees, addPermission, removePermission } = useGrantees(aclUrl);
+  const loadUsersWithPermissions = useCallback(() => {
+    setLoadingGrantees(true);
+    loadGrantees()
+      .then(setGrantees)
+      .catch(() => notification.error('Failed to load grantees list'))
+      .finally(() => setLoadingGrantees(false));
+  }, []);
+
+  const userHasPermission = useCallback(
+    user => (user.id === author.id || !!get(find(grantees, { id: user.id }), 'accessType')),
+    [grantees],
+  );
+
+  useEffect(() => {
+    loadUsersWithPermissions();
+  }, [aclUrl]);
+
+  return (
+    <Modal
+      {...dialog.props}
+      className="permissions-editor-dialog"
+      title={<PermissionsEditorDialogHeader context={context} />}
+      footer={(<Button onClick={dialog.dismiss}>Close</Button>)}
+    >
+      <UserSelect
+        onSelect={userId => addPermission(userId).then(loadUsersWithPermissions)}
+        shouldShowUser={user => !userHasPermission(user)}
+      />
+      <div className="d-flex align-items-center">
+        <h5 className="flex-fill">Users with permissions</h5>
+        {loadingGrantees && <i className="fa fa-spinner fa-pulse" />}
+      </div>
+      <div className="scrollbox p-5 m-t-10" style={{ maxHeight: '40vh' }}>
```

Comment:
A small layout suggestion.
- Give the "d-flex align-items-center" element `m-t-5`.
- Remove the `m-t-10` from scrollbox.